### ISSUE TITLE
Hybrid mask mode: --restore skips candidates

### ISSUE THREAD

**Issue by frank-dittrich** [magnumripper/JohnTheRipper#1859]
```
(bleeding-jumbo)run $ ./john testhashes --format=dynamic='md5($u.$p)' --pot=test.pot --session=test --wordlist=facebook-firstnames.txt --mask='?w-?d?d'
Using default input encoding: UTF-8
Loaded 2000000 password hashes with 2000000 different salts (dynamic=md5($u.$p) [256/256 AVX2 8x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:00 0.00% (ETA: 09:53:54) 0g/s 0p/s 7785Kc/s 7785KC/s michael-11..mary-78
0g 0:00:00:01 0.00% (ETA: 2015-10-30 12:52) 0g/s 0p/s 12858Kc/s 12858KC/s michael-11..mary-78
0g 0:00:00:02 0.00% (ETA: 2015-11-01 15:50) 0g/s 0p/s 14735Kc/s 14735KC/s michael-11..mary-78
0g 0:00:00:03 0.00% (ETA: 2015-11-03 18:49) 0g/s 0p/s 15330Kc/s 15330KC/s michael-11..mary-78
Session aborted
(bleeding-jumbo)run $ ./john --restore=test
Loaded 2000000 password hashes with 2000000 different salts (dynamic=md5($u.$p) [256/256 AVX2 8x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:04 0.00% (ETA: 2015-11-01 16:50) 0g/s 0p/s 14411Kc/s 14411KC/s mary-11
0g 0:00:00:04 0.00% (ETA: 2015-11-01 16:50) 0g/s 0p/s 15381Kc/s 15381KC/s mary-11..bill-78
0g 0:00:00:05 0.00% (ETA: 2015-11-02 18:34) 0g/s 0p/s 15595Kc/s 15595KC/s mary-11..bill-78
0g 0:00:00:06 0.00% (ETA: 2015-11-03 20:18) 0g/s 0p/s 15876Kc/s 15876KC/s mary-11..bill-78
0g 0:00:00:06 0.00% (ETA: 2015-11-03 20:18) 0g/s 0p/s 15876Kc/s 15876KC/s mary-11..bill-78
Session aborted
(bleeding-jumbo)run $ ./john --restore=test
Loaded 2000000 password hashes with 2000000 different salts (dynamic=md5($u.$p) [256/256 AVX2 8x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:07 0.00% (ETA: 2015-11-02 08:39) 0g/s 0p/s 13608Kc/s 13608KC/s bill-11
0g 0:00:00:07 0.00% (ETA: 2015-11-02 08:39) 0g/s 0p/s 13958Kc/s 13958KC/s bill-11..lauren-78
0g 0:00:00:08 0.00% (ETA: 2015-11-03 01:37) 0g/s 0p/s 13091Kc/s 13091KC/s bill-11..lauren-78
0g 0:00:00:09 0.00% (ETA: 2015-11-03 18:35) 0g/s 0p/s 12306Kc/s 12306KC/s bill-11..lauren-78
0g 0:00:00:09 0.00% (ETA: 2015-11-03 18:35) 0g/s 0p/s 12306Kc/s 12306KC/s bill-11..lauren-78
Session aborted
```
Max. keys per crypt is 3360 for this format.
Processing 3360 candidates for 2000000 hashes takes several minutes.
Apparently, when restoring, john continues with the last word from the word list which has been used for the current block, instead of starting with the first input word that has been used for this block.

If the number of candidates which mask mode creates per input word is a multiple of max. keys per crypt (e.g., if max. keys per crypt is 1) or if max. keys per crypt is a multiple of the number of candidates generated per input word , the problem doesn't occur.

**Comment by frank-dittrich** [comment]
With just 20000 hashes (for faster processing...):
```
Loaded 20000 password hashes with 20000 different salts (dynamic=md5($u.$p) [256/256 AVX2 8x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:00 0.00% (ETA: 11:30:39) 0g/s 0p/s 5735Kc/s 5735KC/s michael-11..mary-78
0g 0:00:00:01 0.00% (ETA: 2015-10-30 14:29) 0g/s 0p/s 10154Kc/s 10154KC/s michael-11..mary-78
0g 0:00:00:02 0.00% (ETA: 2015-11-01 17:27) 0g/s 0p/s 11866Kc/s 11866KC/s michael-11..mary-78
0g 0:00:00:03 0.00% (ETA: 2015-11-03 20:25) 0g/s 0p/s 13383Kc/s 13383KC/s michael-11..mary-78
0g 0:00:00:04 0.00% (ETA: 2015-11-05 23:24) 0g/s 0p/s 13699Kc/s 13699KC/s michael-11..mary-78
0g 0:00:00:05 0.00% (ETA: 2015-11-02 18:20) 0g/s 574.3p/s 13399Kc/s 13399KC/s mary-15..katie-70
0g 0:00:00:06 0.00% (ETA: 2015-11-03 19:42) 0g/s 490.5p/s 13724Kc/s 13724KC/s mary-15..katie-70
0g 0:00:00:07 0.00% (ETA: 2015-11-04 21:04) 0g/s 428.0p/s 13182Kc/s 13182KC/s mary-15..katie-70
0g 0:00:00:08 0.00% (ETA: 2015-11-05 22:26) 0g/s 379.2p/s 12945Kc/s 12945KC/s mary-15..katie-70
0g 0:00:00:09 0.00% (ETA: 2015-11-06 23:48) 0g/s 341.1p/s 13210Kc/s 13210KC/s mary-15..katie-70
0g 0:00:00:10 0.00% (ETA: 2015-11-04 11:32) 0g/s 619.3p/s 13442Kc/s 13442KC/s katie-12..tyler-74
0g 0:00:00:11 0.00% (ETA: 2015-11-05 04:21) 0g/s 567.0p/s 13716Kc/s 13716KC/s katie-12..tyler-74
0g 0:00:00:12 0.00% (ETA: 2015-11-05 21:09) 0g/s 522.9p/s 13807Kc/s 13807KC/s katie-12..tyler-74
0g 0:00:00:13 0.00% (ETA: 2015-11-06 13:57) 0g/s 485.1p/s 13407Kc/s 13407KC/s katie-12..tyler-74
0g 0:00:00:14 0.00% (ETA: 2015-11-07 06:45) 0g/s 452.5p/s 13024Kc/s 13024KC/s katie-12..tyler-74
0g 0:00:00:15 0.00% (ETA: 2015-11-05 08:24) 0g/s 635.9p/s 12809Kc/s 12809KC/s tyler-16..emma-73
0g 0:00:00:16 0.00% (ETA: 2015-11-05 20:59) 0g/s 598.2p/s 12568Kc/s 12568KC/s tyler-16..emma-73
0g 0:00:00:17 0.00% (ETA: 2015-11-06 09:35) 0g/s 564.7p/s 12689Kc/s 12689KC/s tyler-16..emma-73
0g 0:00:00:18 0.00% (ETA: 2015-11-06 22:10) 0g/s 532.2p/s 12904Kc/s 12904KC/s tyler-16..emma-73
Session aborted
(bleeding-jumbo)run $ ./john --restore=test
Loaded 20000 password hashes with 20000 different salts (dynamic=md5($u.$p) [256/256 AVX2 8x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:19 0.00% (ETA: 2015-11-05 11:44) 0g/s 530.5p/s 12863Kc/s 12863KC/s emma-74
0g 0:00:00:19 0.00% (ETA: 2015-11-05 11:44) 0g/s 530.5p/s 13158Kc/s 13158KC/s emma-74..pat-63
0g 0:00:00:20 0.00% (ETA: 2015-11-05 21:51) 0g/s 504.0p/s 12872Kc/s 12872KC/s emma-74..pat-63
0g 0:00:00:21 0.00% (ETA: 2015-11-06 07:58) 0g/s 480.0p/s 12658Kc/s 12658KC/s emma-74..pat-63
0g 0:00:00:22 0.00% (ETA: 2015-11-06 18:05) 0g/s 458.1p/s 12690Kc/s 12690KC/s emma-74..pat-63
0g 0:00:00:23 0.00% (ETA: 2015-11-07 04:12) 0g/s 438.2p/s 12873Kc/s 12873KC/s emma-74..pat-63
Session aborted
```

Commit ddad7f91d8a44a9b4c88ee746218507403f95bd6 changed the mask mode logic to use frequency information. This makes it somewhat harder to see what is happening.
That's why I retried this after a `git revert ddad7f91d8a44a9b4c88ee746218507403f95bd6`.

```
(bleeding-jumbo)run $ rm test.*;./john testhashes --format=dynamic='md5($u.$p)' --pot=test.pot --session=test --wordlist=facebook-firstnames.txt --mask='?w-?d?d'
Using default input encoding: UTF-8
Loaded 20000 password hashes with 20000 different salts (dynamic=md5($u.$p) [256/256 AVX2 8x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:00 0.00% (ETA: 11:53:12) 0g/s 0p/s 8705Kc/s 8705KC/s michael-00..mary-95
0g 0:00:00:01 0.00% (ETA: 2015-10-30 14:51) 0g/s 0p/s 13031Kc/s 13031KC/s michael-00..mary-95
0g 0:00:00:02 0.00% (ETA: 2015-11-01 17:50) 0g/s 0p/s 13041Kc/s 13041KC/s michael-00..mary-95
Session aborted
(bleeding-jumbo)run $ ./john --restore=test
Loaded 20000 password hashes with 20000 different salts (dynamic=md5($u.$p) [256/256 AVX2 8x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:03 0.00% (ETA: 2015-10-31 17:05) 0g/s 0p/s 12433Kc/s 12433KC/s mary-00
0g 0:00:00:03 0.00% (ETA: 2015-10-31 17:05) 0g/s 0p/s 14091Kc/s 14091KC/s mary-00..bill-95
0g 0:00:00:04 0.00% (ETA: 2015-11-01 18:49) 0g/s 0p/s 13203Kc/s 13203KC/s mary-00..bill-95
0g 0:00:00:05 0.00% (ETA: 2015-11-02 20:33) 0g/s 0p/s 13843Kc/s 13843KC/s mary-00..bill-95
0g 0:00:00:06 0.00% (ETA: 2015-11-03 22:17) 0g/s 0p/s 13484Kc/s 13484KC/s mary-00..bill-95
0g 0:00:00:07 0.00% (ETA: 2015-11-05 00:01) 0g/s 0p/s 13736Kc/s 13736KC/s mary-00..bill-95
0g 0:00:00:09 0.00% (ETA: 2015-11-03 19:07) 0g/s 373.3p/s 12573Kc/s 12573KC/s bill-06..tyler-91
0g 0:00:00:11 0.00% (ETA: 2015-11-05 04:43) 0g/s 305.4p/s 12838Kc/s 12838KC/s bill-06..tyler-91
Session aborted
(bleeding-jumbo)run $ ./john --restore=test
Loaded 20000 password hashes with 20000 different salts (dynamic=md5($u.$p) [256/256 AVX2 8x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:12 0.00% (ETA: 2015-11-03 19:54) 0g/s 280.0p/s 11768Kc/s 11768KC/s tyler-00
0g 0:00:00:12 0.00% (ETA: 2015-11-03 19:54) 0g/s 280.0p/s 12173Kc/s 12173KC/s tyler-00..larry-95
0g 0:00:00:13 0.00% (ETA: 2015-11-04 08:34) 0g/s 258.4p/s 12086Kc/s 12086KC/s tyler-00..larry-95
0g 0:00:00:14 0.00% (ETA: 2015-11-04 21:14) 0g/s 240.0p/s 12199Kc/s 12199KC/s tyler-00..larry-95
0g 0:00:00:14 0.00% (ETA: 2015-11-04 21:14) 0g/s 240.0p/s 12199Kc/s 12199KC/s tyler-00..larry-95
Session aborted
```

**Comment by frank-dittrich** [comment]
The same problem exists if --mask is combined with --incremental (and probably other modes) instead of --wordlist:
```
(bleeding-jumbo)run $ rm test.*;./john testhashes --format=dynamic='md5($u.$p)' --pot=test.pot --session=test --incremental --min-len=8 --max-len=8 --mask='?w-?d?d'
Using default input encoding: UTF-8
Loaded 20000 password hashes with 20000 different salts (dynamic=md5($u.$p) [256/256 AVX2 8x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:00  0g/s 0p/s 7883Kc/s 7883KC/s 12345-00..14324-95
0g 0:00:00:01  0g/s 0p/s 12218Kc/s 12218KC/s 12345-00..14324-95
0g 0:00:00:02  0g/s 0p/s 13157Kc/s 13157KC/s 12345-00..14324-95
0g 0:00:00:03  0g/s 0p/s 13941Kc/s 13941KC/s 12345-00..14324-95
0g 0:00:00:04  0g/s 722.5p/s 14517Kc/s 14517KC/s 14324-06..mishy-91
0g 0:00:00:05  0g/s 592.5p/s 14226Kc/s 14226KC/s 14324-06..mishy-91
0g 0:00:00:06  0g/s 505.2p/s 13119Kc/s 13119KC/s 14324-06..mishy-91
0g 0:00:00:07  0g/s 439.2p/s 13267Kc/s 13267KC/s 14324-06..mishy-91
0g 0:00:00:07  0g/s 430.7p/s 13012Kc/s 13012KC/s 14324-06..mishy-91
Session aborted
(bleeding-jumbo)run $ ./john --restore=test
Loaded 20000 password hashes with 20000 different salts (dynamic=md5($u.$p) [256/256 AVX2 8x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:08  0g/s 420.0p/s 12687Kc/s 12687KC/s misse-95
0g 0:00:00:08  0g/s 420.0p/s 13402Kc/s 13402KC/s misse-95..miler-81
0g 0:00:00:09  0g/s 373.3p/s 13544Kc/s 13544KC/s misse-95..miler-81
0g 0:00:00:10  0g/s 336.0p/s 13657Kc/s 13657KC/s misse-95..miler-81
0g 0:00:00:11  0g/s 305.4p/s 13763Kc/s 13763KC/s misse-95..miler-81
0g 0:00:00:11  0g/s 305.4p/s 13763Kc/s 13763KC/s misse-95..miler-81
Session aborted
```

**Comment by magnumripper** [comment]
Bad bug, good catch. When I audited this some time ago I was mainly using `-stdout` which didn't show the problem (lesson learned). I'm near certain it's all good now: First because I now understand what we did wrong, and second because I verified the fix by adding `log_event()` debug lines in `set_key()` and `crypt_all()` of a slow format. I could confirm that the restore point is now never past the last `crypt_all()` call.

I'm pretty sure we're now restoring hybrid mask "optimally" at all times, that is, neither too early nor too late.

**Comment by frank-dittrich** [comment]
@magnumripper Thanks for fixing it.

Do we need to enhance the --restore test in jtrts.pl?
Should we announce on john-users that we had this bug, and that it is now fixed?

**Comment by magnumripper** [comment]
Do we have any restore test in TS at all yet?

I'm not sure we need to announce it. Typically the missed range would be miniscule compared to total keyspace. Also, if you have older code running, you can't resume with this newer one anyway since we changed to frequency based ordering.

**Comment by frank-dittrich** [comment]
From `./jtrts.pl -help` output:
```
    -restore       Runs a single test, in -restore mode. To do this, we run
                   john a few seconds, then resume this run until we have
                   completed the work, then check that all are properly cracked
```

**Comment by jfoug** [comment]
I was under the assumption that a single type of restore would be adequate.  Obviously I was not correct in that assumption.  A single format 'type' probably should be adequate, but we need to build a restore test for each input type.

wordlist
single
wordlist - rules
mask
markov
prince
rexgen
rexgen $0  (wordlist)
.... others ? ....

I think a wordlist is all that is tested today.

**Comment by jfoug** [comment]
Also, the current restore is still sort of POC. it works, but is not super robust.

**Comment by magnumripper** [comment]
Wordlist mode is one type of restore, mask mode is another, wordlist + mask is a third. However, if we test all modes you list and then (any) one of them (eg. wordlist) with hybrid mask, I'm pretty sure we're set. We should not need to verify the hybrid mask restore for each and every parent mode because if one works, all should be fine.

**Comment by jfoug** [comment]
I added a list of modes in that TS issue.  I also added that we may need to restore test -fork.   Would that test need to be for all modes?

On thing I have seen on -fork, is that each fork can end up working on different rules during the run, i.e. some get ahead of the others.  I do not know if that matters at all (I would think not), it is just an observation.  I have 5 million salted 'fast' hashes running with about 100k rules and rock-u as the word list.  I see a pretty large divergence between the threads.

**Comment by frank-dittrich** [comment]
As long as your restore test also makes use of -passthru when starting john for the first time, there's no need to add special tests with --fork.

**Comment by magnumripper** [comment]
I agree with Frank, we can run it with fork when we want to.

Diverging processes are totally natural. Both for MPI, fork and manual "-node", the processes run 100% detached from each other - no barriers, no syncing whatsoever.

**Comment by jfoug** [comment]
This is still a problem.

**Comment by magnumripper** [comment]
I'm sure the TS is correct about it, but I really can't see what is wrong now. Since ddb63a1 it does it the "core way", that is, cracker.c decides when it's safe to call `fix_state()`. The good news is we now have a good test method - I guess I'll just dive into it again.

**Comment by jfoug** [comment]
Actually, on my VM it WAS working, until I pulled code ....   Go figure, lol.

**Comment by magnumripper** [comment]
My current theory is that the session store is actually totally correct but the *resume* of it is not! I didn't even think about that until now.

**Comment by magnumripper** [comment]
Manual run (using `?w[0-9][0-9]` instead of `?w?d?d` to get natural order, and using `-max-run=120` so to only get one stop/go cycle.

tail of first half:
```
vzlht eifja70    (u1414)
altiuelznvb69    (u1393)
vzlht eifja60    (u1214)
altiuelznvb59    (u1193)
vzlht eifja50    (u1014)
altiuelznvb49    (u993)
vzlht eifja40    (u814)
altiuelznvb39    (u793)
vzlht eifja30    (u614)
altiuelznvb29    (u593)
vzlht eifja20    (u414)
vzlht eifja10    (u214)
vzlht eifja00    (u14)
vzlht eifja91    (u1834)
vzlht eifja90    (u1814)
vzlht eifja81    (u1634)
vzlht eifja80    (u1614)
vzlht eifja71    (u1434)
vzlht eifja61    (u1234)
vzlht eifja51    (u1034)
vzlht eifja41    (u834)
vzlht eifja32    (u654)
vzlht eifja31    (u634)
vzlht eifja22    (u454)
vzlht eifja21    (u434)
vzlht eifja12    (u254)
vzlht eifja11    (u234)
1422g 0:00:01:58 75.86% (ETA: 20:28:51) 11.96g/s 11.84p/s 15524c/s 15524C/s vzlht eifja80..vzlht eifja32
Use the "--show" option to display all of the cracked passwords reliably
Session stopped (max run-time reached)
```

rec file when stopped:
```
$ cat tst-.rec 
REC4
10
--session=tst-
--mask=?w[0-9][0-9]
--wordlist=bitcoin_restart_rules_tst.dic
bitcoin_restart_rules_tst.in
--pot=tst-.pot
--max-run-time=120
--format=bitcoin
--input-encoding=ASCII
--target-encoding=ASCII
119
1422
1c26e0
0
0
1c26e0
0
580
0
0
0
75
0
0
0
15
101
2
0
7
0
```

head of second half:
```
Warning: check for duplicates partially bypassed to speedup loading
Loaded 2000 password hashes with 2000 different salts (Bitcoin [SHA512 AES 128/128 AVX 2x])
Remaining 578 password hashes with 578 different salts
Will run 8 OpenMP threads
Press 'q' or Ctrl-C to abort, almost any other key for status
1422g 0:00:01:59 80.79% (ETA: 20:32:23) 11.94g/s 11.83p/s 15503c/s 15503C/s aloiaytwe70
aloiaytwe91      (u1835)
aloiaytwe90      (u1815)
aloiaytwe81      (u1635)
aloiaytwe80      (u1615)
aloiaytwe71      (u1435)
aloiaytwe70      (u1415)
aloiaytwe61      (u1235)
aloiaytwe51      (u1035)
aloiaytwe41      (u835)
aloiaytwe31      (u635)
aloiaytwe22      (u455)
aloiaytwe21      (u435)
aloiaytwe12      (u255)
aloiaytwe11      (u235)
aloiaytwe02      (u55)
aloiaytwe01      (u35)
aloiaytwe92      (u1855)
aloiaytwe83      (u1675)
aloiaytwe82      (u1655)
aloiaytwe73      (u1475)
aloiaytwe72      (u1455)
aloiaytwe63      (u1275)
aloiaytwe62      (u1255)
aloiaytwe53      (u1075)
aloiaytwe52      (u1055)
aloiaytwe43      (u875)
aloiaytwe42      (u855)
aloiaytwe33      (u675)
aloiaytwe32      (u655)
aloiaytwe23      (u475)
aloiaytwe13      (u275)
aloiaytwe03      (u75)
aloiaytwe94      (u1895)
aloiaytwe93      (u1875)
aloiaytwe84      (u1695)
aloiaytwe74      (u1495)
aloiaytwe64      (u1295)
aloiaytwe54      (u1095)
aloiaytwe45      (u915)
aloiaytwe44      (u895)
aloiaytwe35      (u715)
aloiaytwe34      (u695)
aloiaytwe25      (u515)
aloiaytwe24      (u495)
aloiaytwe15      (u315)
aloiaytwe14      (u295)
aloiaytwe05      (u115)
aloiaytwe04      (u95)
aloiaytwe96      (u1935)
aloiaytwe95      (u1915)
aloiaytwe86      (u1735)
aloiaytwe85      (u1715)
aloiaytwe76      (u1535)
aloiaytwe75      (u1515)
aloiaytwe66      (u1335)
aloiaytwe65      (u1315)
aloiaytwe56      (u1135)
aloiaytwe55      (u1115)
aloiaytwe46      (u935)
aloiaytwe36      (u735)
aloiaytwe26      (u535)
aloiaytwe16      (u335)
aloiaytwe07      (u155)
aloiaytwe06      (u135)
aloiaytwe97      (u1955)
aloiaytwe87      (u1755)
aloiaytwe77      (u1555)
aloiaytwe68      (u1375)
aloiaytwe67      (u1355)
aloiaytwe58      (u1175)
aloiaytwe57      (u1155)
aloiaytwe48      (u975)
aloiaytwe47      (u955)
aloiaytwe38      (u775)
aloiaytwe37      (u755)
aloiaytwe28      (u575)
aloiaytwe27      (u555)
aloiaytwe18      (u375)
aloiaytwe17      (u355)
aloiaytwe08      (u175)
aloiaytwe99      (u1995)
aloiaytwe98      (u1975)
aloiaytwe89      (u1795)
aloiaytwe88      (u1775)
aloiaytwe79      (u1595)
aloiaytwe78      (u1575)
aloiaytwe69      (u1395)
aloiaytwe59      (u1195)
aloiaytwe49      (u995)
aloiaytwe39      (u795)
aloiaytwe29      (u595)
4890bjsi20       (u416)
aloiaytwe19      (u395)
4890bjsi10       (u216)
aloiaytwe09      (u195)
```

**Comment by magnumripper** [comment]
```
$ ../run/john bitcoin_restart_rules_tst.in -pot=tst-.pot -form=bitcoin -show=left > left
Warning: check for duplicates partially bypassed to speedup loading

1915 password hashes cracked, 85 left
```
```
$ grep -Ff left bitcoin_restart_rules_tst.in | cut -d: -f5 | sort
aloiaytwe00
aloiaytwe10
aloiaytwe20
aloiaytwe30
aloiaytwe40
aloiaytwe50
aloiaytwe60
vzlht eifja01
vzlht eifja02
vzlht eifja03
vzlht eifja04
vzlht eifja05
vzlht eifja06
vzlht eifja07
vzlht eifja08
vzlht eifja09
vzlht eifja13
vzlht eifja14
vzlht eifja15
vzlht eifja16
vzlht eifja17
vzlht eifja18
vzlht eifja19
vzlht eifja23
vzlht eifja24
vzlht eifja25
vzlht eifja26
vzlht eifja27
vzlht eifja28
vzlht eifja29
vzlht eifja33
vzlht eifja34
vzlht eifja35
vzlht eifja36
vzlht eifja37
vzlht eifja38
vzlht eifja39
vzlht eifja42
vzlht eifja43
vzlht eifja44
vzlht eifja45
vzlht eifja46
vzlht eifja47
vzlht eifja48
vzlht eifja49
vzlht eifja52
vzlht eifja53
vzlht eifja54
vzlht eifja55
vzlht eifja56
vzlht eifja57
vzlht eifja58
vzlht eifja59
vzlht eifja62
vzlht eifja63
vzlht eifja64
vzlht eifja65
vzlht eifja66
vzlht eifja67
vzlht eifja68
vzlht eifja69
vzlht eifja72
vzlht eifja73
vzlht eifja74
vzlht eifja75
vzlht eifja76
vzlht eifja77
vzlht eifja78
vzlht eifja79
vzlht eifja82
vzlht eifja83
vzlht eifja84
vzlht eifja85
vzlht eifja86
vzlht eifja87
vzlht eifja88
vzlht eifja89
vzlht eifja92
vzlht eifja93
vzlht eifja94
vzlht eifja95
vzlht eifja96
vzlht eifja97
vzlht eifja98
vzlht eifja99
```

**Comment by magnumripper** [comment]
> Actually, on my VM it WAS working, until I pulled code .... Go figure, lol.

Interesting. Please do a `git reflog | head` on that VM and post here - maybe we can see at what commit it did work.

**Comment by jfoug** [comment]
[ghost@localhost src]$ git reflog | head
612d3eb HEAD@{0}: pull: Fast-forward
2fe6b2e HEAD@{1}: pull: Fast-forward
cc20c79 HEAD@{2}: pull: Fast-forward
78ee669 HEAD@{3}: pull: Fast-forward
58a4f66 HEAD@{4}: pull: Fast-forward
9127926 HEAD@{5}: pull: Fast-forward
07f71fb HEAD@{6}: pull: Fast-forward
d5d4edb HEAD@{7}: pull: Fast-forward
b558305 HEAD@{8}: pull: Fast-forward
045f71d HEAD@{9}: pull: Fast-forward

**Comment by magnumripper** [comment]
2fe6b2e was indeed before fixing the bug @frank-dittrich reported. So TS sees the problem after we fixed that but not before? That's backwards, I need to try that..

**Comment by magnumripper** [comment]
What the actual f@ck. It works fine if I revert ddb63a1. How can that be?

**Comment by magnumripper** [comment]
It even works when also adding `-mkpc=11` (which is a prime).

**Comment by jfoug** [comment]
All I can do is 'report' results. I did not follow along with the initial bug report/fix until the TS poc -resume issue came up.

But I saw same thing.  A build on the VM which was not updated 'worked', but then when I pulled to get current code, things started having problems.  Go figure....

**Comment by magnumripper** [comment]
I'm on it. The change in ddb63a1 is basically correct but there's some (other) issue now, with the parent mode.

I believe one bug was definitely fixed in ddb63a1 (which TS unfortunately does not detect) and we see *another* bug now (which it does detect). I wish I could understand why it doesn't see any bug prior to it. It might be "luck" with multiples of things, somehow.

**Comment by jfoug** [comment]
is it possible the 'bug' you saw was not a bug, but correct behavior?   Could it be that the group of data IS properly fully processed prior to the exit, or possibly it is 'fully' completed on a -max-time=xxx run.    It may be that we need to fire up a thread in the perl script and issue a kill -9 against the pid of the running john, so that we violently kill it, and then resume from there (or would that not drop a restore file???)

**Comment by magnumripper** [comment]
This is hard to fully grasp. We have two different problem situations:
1. We must not fix *any* state later than what is actually processed and done with *all salts*. This is taken care of by cracker.c - it only calls fix_state when appropriate.
2. We must not fix *parent* state later than what is fully processed with *all masks*

I think the bug was that was fixed addressed # 1 (prior to that commit **we** called parent fix_state at an inappropriate time) but introduced # 2.

**Comment by magnumripper** [comment]
Here's current behavior:
```
11111007         (u140)
11111006         (u120)
mask_fix_state()
11111191         (u1821)
11111087         (u1740)
11111181         (u1621)
11111077         (u1540)
11111171         (u1421)
11111067         (u1340)
11111161         (u1221)
11111057         (u1140)
11111151         (u1021)
11111047         (u940)
11111141         (u821)
11111131         (u621)
11111121         (u421)
11111111         (u221)
11111110         (u201)
11111101         (u21)
mask_fix_state()
fix_state()
11111100         (u1)
11111192         (u1841)
11111190         (u1801)
11111182         (u1641)
11111180         (u1601)
11111170         (u1401)
11111160         (u1201)
11111152         (u1041)
11111150         (u1001)
11111140         (u801)
11111132         (u641)
11111130         (u601)
11111122         (u441)
11111120         (u401)
11111112         (u241)
11111102         (u41)
mask_fix_state()
11111193         (u1861)
11111183         (u1661)
11111173         (u1461)
11111172         (u1441)
11111163         (u1261)
11111162         (u1241)
11111153         (u1061)
134g 0:00:00:17 10.00% (ETA: 14:10:13) 7.882g/s 7.529p/s 16085c/s 16085C/s 11111142..11111129
Use the "--show" option to display all of the cracked passwords reliably
Session stopped (max run-time reached)
```
We clearly see that we call parent's fix_state() at just the right time. However,
```
$ ../run/john -res=tst-
Warning: check for duplicates partially bypassed to speedup loading
Loaded 2000 password hashes with 2000 different salts (Bitcoin [SHA512 AES 128/128 AVX 2x])
Remaining 1866 password hashes with 1866 different salts
Will run 8 OpenMP threads
fix_state()
mask_fix_state()
Press 'q' or Ctrl-C to abort, almost any other key for status
134g 0:00:00:18 15.00% (ETA: 14:09:35) 7.444g/s 7.111p/s 15192c/s 15192C/s 11111252
11111293         (u1862)
11111283         (u1662)
(...)
```
...despite that, it picks up **next** parent word. I think the problem is in wordlist.c - but why does it not happen with pure wordlist or wordlist+rules?

**Comment by magnumripper** [comment]
Confirmed! Hybrid `-inc -mask` works like a champ. The problem is in wordlist.c

**Comment by magnumripper** [comment]
I tested `-w -ru -mask:?w?d?d` and that one has problem of suboptimal resuming. It re-tries a good number of candidates. That's much less of a problem (better safe than sorry) but should be avoided. After a couple of resumes (30 s runs) it got to the point it made no progress in one run. I bumped max run time so it ran to finish. In the end it obviously cracked all of them.

**Comment by frank-dittrich** [comment]
I think I saw something similar with --prince mode (no rules), but don't have time right now to verify this.

**Comment by magnumripper** [comment]
The wordlist+rules+mask problem is not caused by hybrid mask at all, it's a property of wordlist+rules. But mask mode multiplies the consequences...

**Comment by jfoug** [comment]
I have known there is a 'over conservative' problem in the -w + -ru mode for a while.  It REALLY shows up on tests like the AM bcrypt stuff, which is SUPER slow.

**Comment by magnumripper** [comment]
I think we need to establish whether core wordlist+rules are the same.

**Comment by magnumripper** [comment]
Fixed now. I had to add another level of fix_state. After a batch of mask or regex, we record the wordlist state but only to another set of variables. At next proper time (as deemed by cracker.c) the real fix_state is called - and will use those recorded values.

Theoretically this should be needed for other modes too (so we should tweak the fix_state interface globally). I'm not sure why it's not. Perhaps all of them resume early. I have a slightly bad feeling about it - perhaps there are edge cases where we'll actually miss something.
